src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -107,17 +107,63 @@ private static class Compiler<TModel, TResult>
                 // by around one microsecond, so it's not worth it to complicate the logic here with
                 // an architecture check.
 
-                var memberExpression = expression.Body as MemberExpression;
-                if (memberExpression != null)
+                if (expression.Body is MemberExpression memberExpression)
                 {
-                    if (memberExpression.Expression == expression.Parameters[0] || memberExpression.Expression == null)
+                    if (memberExpression.Expression == null)
                     {
-                        // model => model.Member or model => StaticMember
+                        // model => model => StaticMember
                         return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
                     }
 
-                    var constantExpression = memberExpression.Expression as ConstantExpression;
-                    if (constantExpression != null)
+                    if (memberExpression.Expression == expression.Parameters[0])
+                    {
+                        // model => model.Member
+                        if (memberExpression.Member.DeclaringType.IsValueType &&
+                            Nullable.GetUnderlyingType(memberExpression.Type) == null)
+                        {
+                            // struct.Member
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ => expression.Compile());
+                        }
+                        else if (!memberExpression.Type.IsValueType &&
+                                 Nullable.GetUnderlyingType(memberExpression.Type) != null)
+                        {
+                            // class.class, class.nullable, nullable.class, nullable.nullable
+                            return _simpleMemberAccessCache.GetOrAdd(memberExpression.Member, _ =>
+                            {
+                                // Add null conditional
+                                var replacementMemberExpression = Expression.Lambda<Func<TModel, TResult>>(
+                                    Expression.Condition(
+                                        Expression.NotEqual(memberExpression.Expression,
```

Comment:
FYI this is https://msdn.microsoft.com/en-us/library/bb354591(v=vs.110).aspx - which generates an `==` and does overload resolution via reflection when this is compiled. It might be better to use `object.ReferenceEquals` instead here. 

I think this is probably fine since we're dealing with object, but if we wanted more precise control this may give it.